README.md
# Unofficial Learncube rest api client


CHANGELOG.md
# v0.1.2 (Fri Nov 05 2021)

#### ⚠️ Pushed to `master`

- Publish to github registry (ja@krzysztofbukowski.pl)

#### Authors: 1

- Krzysztof Bukowski (ja@krzysztofbukowski.pl)

---

# v0.1.1 (Fri Nov 05 2021)

#### ⚠️ Pushed to `master`

- Fixed env (ja@krzysztofbukowski.pl)
- Fixed env vars (ja@krzysztofbukowski.pl)
- Added .npmrc (ja@krzysztofbukowski.pl)
- Added files section in package.json (ja@krzysztofbukowski.pl)
- Configured auto (ja@krzysztofbukowski.pl)

#### Authors: 1

- Krzysztof Bukowski (ja@krzysztofbukowski.pl)

---

# v0.1.1 (Fri Nov 05 2021)

#### ⚠️ Pushed to `master`

- Fixed env (ja@krzysztofbukowski.pl)
- Fixed env vars (ja@krzysztofbukowski.pl)
- Added .npmrc (ja@krzysztofbukowski.pl)
- Added files section in package.json (ja@krzysztofbukowski.pl)
- Configured auto (ja@krzysztofbukowski.pl)

#### Authors: 1

- Krzysztof Bukowski (ja@krzysztofbukowski.pl)

---

# v0.1.1 (Fri Nov 05 2021)

#### ⚠️ Pushed to `master`

- Fixed env (ja@krzysztofbukowski.pl)
- Fixed env vars (ja@krzysztofbukowski.pl)
- Added .npmrc (ja@krzysztofbukowski.pl)
- Added files section in package.json (ja@krzysztofbukowski.pl)
- Configured auto (ja@krzysztofbukowski.pl)

#### Authors: 1

- Krzysztof Bukowski (ja@krzysztofbukowski.pl)


src/client.test.ts
import fetch from 'node-fetch';
import { del, get, post, put, request } from './client';

jest.mock('node-fetch');

beforeEach(() => {
  ((fetch as unknown) as jest.Mock).mockClear();
});

describe('client', () => {
  describe('request', () => {
    it('sends authorization header if present', async () => {
      ((fetch as unknown) as jest.Mock).mockResolvedValue({
        status: 200,
        json: jest.fn(),
      });

      await request(
        'GET',
        'https://api.learncube.com/virtual-classroom/',
        {},
        'some-token',
      );

      expect(fetch).toHaveBeenCalledWith(
        'https://api.learncube.com/virtual-classroom/',
        expect.objectContaining({
          headers: expect.objectContaining({
            Authorization: 'Bearer some-token',
          }),
        }),
      );
    });
  });

  it('get should send a request with query params', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValue({
      status: 200,
      json: jest.fn().mockResolvedValue({
        someData: 10,
      }),
    });

    const response = await get('https://api.learncube.com/virtual-classroom/', {
      page: '1',
      limit: '10',
    });

    expect(fetch).toHaveBeenCalledWith(
      'https://api.learncube.com/virtual-classroom/?page=1&limit=10',
      expect.objectContaining({
        body: undefined,
      }),
    );

    expect(response).toMatchInlineSnapshot(`
      Object {
        "body": Object {
          "someData": 10,
        },
        "status": 200,
      }
    `);
  });

  it('post should send a request with body params', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValue({
      status: 200,
      json: jest.fn().mockResolvedValue({
        someData: 10,
      }),
    });
    const params = {
      page: '1',
      limit: '10',
    };
    const response = await post(
      'https://api.learncube.com/virtual-classroom/',
      params,
    );

    expect(fetch).toHaveBeenCalledWith(
      'https://api.learncube.com/virtual-classroom/',
      expect.objectContaining({
        method: 'POST',
        body: JSON.stringify(params),
      }),
    );

    expect(response).toMatchInlineSnapshot(`
      Object {
        "body": Object {
          "someData": 10,
        },
        "status": 200,
      }
    `);
  });

  it('put should send a request with body params', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValue({
      status: 200,
      json: jest.fn(),
    });
    const params = {
      page: '1',
      limit: '10',
    };
    const response = await put(
      'https://api.learncube.com/virtual-classroom/',
      params,
    );

    expect(fetch).toHaveBeenCalledWith(
      'https://api.learncube.com/virtual-classroom/',
      expect.objectContaining({
        method: 'PUT',
        body: JSON.stringify(params),
      }),
    );

    expect(response).toMatchInlineSnapshot(`
      Object {
        "body": undefined,
        "status": 200,
      }
    `);
  });

  it('delete should send a request with body params', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValue({
      status: 204,
      json: jest.fn(),
    });

    const response = await del(
      'https://api.learncube.com/virtual-classroom/10',
      undefined,
    );

    expect(fetch).toHaveBeenCalledWith(
      'https://api.learncube.com/virtual-classroom/10',
      expect.objectContaining({
        method: 'DELETE',
        body: undefined,
      }),
    );

    expect(response).toMatchInlineSnapshot(`
      Object {
        "body": undefined,
        "status": 204,
      }
    `);
  });
});


src/authentication.test.ts
import fetch from 'node-fetch';
import { getApiToken, GetApiTokenResponse } from './authentication';

jest.mock('node-fetch');

beforeEach(() => {
  ((fetch as unknown) as jest.Mock).mockClear();
});

describe('authentication', () => {
  describe('get-api-token', () => {
    it('Makes a request and returns client response', async () => {
      ((fetch as unknown) as jest.Mock).mockResolvedValue({
        status: 200,
        json: jest.fn().mockResolvedValue({
          token: 'random-token',
        } as GetApiTokenResponse),
      });

      const result = await getApiToken('public-key', 'private-key');

      expect(result).toMatchInlineSnapshot(`
        Object {
          "body": Object {
            "token": "random-token",
          },
          "status": 200,
        }
      `);

      expect(fetch).toHaveBeenCalledWith(
        expect.stringContaining(
          'https://app.learncube.com/api/virtual-classroom/get-api-token/',
        ),
        expect.objectContaining({
          body: expect.any(String),
        }),
      );
    });
  });
});


src/virtual-classroom/classrooms.test.ts
import fetch from 'node-fetch';
import { StatusCodes } from 'http-status-codes';
import { create, list, read, update, del } from './classrooms';
jest.mock('node-fetch');

beforeEach(() => {
  ((fetch as unknown) as jest.Mock).mockClear();
});

describe('classrooms', () => {
  it('list sends get request', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValue({
      status: StatusCodes.OK,
      json: jest.fn().mockResolvedValue({}),
    });

    const response = await list(
      {
        page: 1,
        teacher_attended: true,
      },
      'api-token',
    );

    expect(fetch).toHaveBeenCalledWith(
      'https://app.learncube.com/api/virtual-classroom/classrooms?page=1&teacher_attended=true',
      expect.objectContaining({
        method: 'GET',
        body: undefined,
        headers: expect.any(Object),
      }),
    );
    expect(response).toMatchInlineSnapshot(`
      Object {
        "body": Object {},
        "status": 200,
      }
    `);
  });

  it('create sends post request', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValue({
      status: StatusCodes.CREATED,
      json: jest.fn().mockResolvedValue({}),
    });

    const params = {
      room_token: 'test-room-token',
    };

    const response = await create(params, 'api-token');

    expect(fetch).toHaveBeenCalledWith(
      'https://app.learncube.com/api/virtual-classroom/classrooms',
      expect.objectContaining({
        method: 'POST',
        body: JSON.stringify(params),
        headers: expect.any(Object),
      }),
    );
    expect(response).toMatchInlineSnapshot(`
      Object {
        "body": Object {},
        "status": 201,
      }
    `);
  });

  it('read sends get request', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValue({
      status: 200,
      json: jest.fn().mockResolvedValue({}),
    });

    const response = await read('uuid-12345', 'api-token');

    expect(fetch).toHaveBeenCalledWith(
      'https://app.learncube.com/api/virtual-classroom/classrooms/uuid-12345',
      expect.objectContaining({
        method: 'GET',
        body: undefined,
        headers: expect.any(Object),
      }),
    );
    expect(response).toMatchInlineSnapshot(`
      Object {
        "body": Object {},
        "status": 200,
      }
    `);
  });

  it('update sends put request', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValue({
      status: StatusCodes.ACCEPTED,
      json: jest.fn().mockResolvedValue({}),
    });

    const params = {
      room_token: 'test-room-token',
    };

    const response = await update('uuid-12345', params, 'api-token');

    expect(fetch).toHaveBeenCalledWith(
      'https://app.learncube.com/api/virtual-classroom/classrooms/uuid-12345',
      expect.objectContaining({
        method: 'PUT',
        body: JSON.stringify(params),
        headers: expect.any(Object),
      }),
    );
    expect(response).toMatchInlineSnapshot(`
      Object {
        "body": Object {},
        "status": 202,
      }
    `);
  });

  it('del sends delete request and returns undefined', async () => {
    ((fetch as unknown) as jest.Mock).mockResolvedValue({
      status: 204,
      json: jest.fn().mockResolvedValue(undefined),
    });

    const response = await del('uuid-12345', 'api-token');

    expect(fetch).toHaveBeenCalledWith(
      'https://app.learncube.com/api/virtual-classroom/classrooms/uuid-12345',
      expect.objectContaining({
        method: 'DELETE',
        body: undefined,
        headers: expect.any(Object),
      }),
    );
    expect(response).toMatchInlineSnapshot(`
      Object {
        "body": undefined,
        "status": 204,
      }
    `);
  });
});


src/virtual-classroom/classrooms.ts
import * as client from '../client';
import { ClientResponse } from '../client';
import { CLASSROOMS_ENDPOINT } from '../endpoints';
import { ListResponse } from './common';

export interface Classroom {
  uuid: string;
  teacher_first_name: string;
  teacher_last_name: string;
  teacher_avatar: string;
  creation_date: string;
  room_token: string | null;
  breakout_parent_token: string | null;
  cancelled: boolean;
  cancelled_by: string | null;
  description: string;
  start: string;
  end: string;
  max_participants: number;
  audio_only: boolean;
  whiteboard_only: boolean;
  teacher_attended: boolean;
  teacher_attended_time: string | null;
  actual_class_start: string | null;
  actual_class_end: string | null;
  actual_duration: number;
  return_url: string | null;
  room_url: string;
  room_review_url: string;
  recorded_class_url: string;
  slug: string;
}

export type ClassroomListResponse = ListResponse<Classroom>;
export type ClassroomListRequestParams = Partial<{
  page: number;
  room_token: string;
  start_before: string;
  start_after: string;
  company_slug: string;
  teacher_attended: boolean;
  teacher_id: string;
}>;

export type ClassroomCreateResponse = Classroom;
export interface ClassroomCreateRequestParams {
  room_token: string;
  cancelled?: boolean;
  description?: string;
  start?: string;
  end?: string;
  max_participants?: number;
  audio_only?: boolean;
  whiteboard_only?: boolean;
  return_url?: string;
  record_class?: boolean;
}

export interface ClassroomUpdateRequestParams {
  room_token: string;
  cancelled?: boolean;
  description?: string;
  start?: string;
  end?: string;
  max_participants?: number;
  audio_only?: boolean;
  whiteboard_only?: boolean;
  return_url?: string;
  record_class?: boolean;
}

export type ClassroomReadResponse = Classroom;
export type ClassroomUpdateResponse = Classroom;
export type ClassroomDeleteResponse = undefined;

export async function list(
  params: ClassroomListRequestParams,
  token: string,
): Promise<ClientResponse<ClassroomListResponse>> {
  return client.get(CLASSROOMS_ENDPOINT, params, token);
}

export async function create(
  params: ClassroomCreateRequestParams,
  token: string,
): Promise<ClientResponse<ClassroomCreateResponse>> {
  return client.post(CLASSROOMS_ENDPOINT, params, token);
}

export async function read(
  uuid: string,
  token: string,
): Promise<ClientResponse<ClassroomReadResponse>> {
  return client.get(`${CLASSROOMS_ENDPOINT}/${uuid}`, undefined, token);
}

export async function update(
  uuid: string,
  params: ClassroomCreateRequestParams,
  token: string,
): Promise<ClientResponse<ClassroomUpdateResponse>> {
  return client.put(`${CLASSROOMS_ENDPOINT}/${uuid}`, params, token);
}

export async function del(
  uuid: string,
  token: string,
): Promise<ClientResponse<ClassroomDeleteResponse>> {
  return client.del(`${CLASSROOMS_ENDPOINT}/${uuid}`, undefined, token);
}


src/virtual-classroom/participants.ts
import { ListResponse } from './common';
import * as client from '../client';
import { PARTICIPANTS_ENDPOINT } from '../endpoints';
import { ClientResponse } from '../client';

export interface Participant {
  uuid: string;
  event_uuid: string;
  room_token: string | null;
  room_slug: string;
  company_slug: string;
  teacher: string;
  user_role: string;
  userid: string | null;
  name: string;
  username: string;
  avatar: string;
  cancelled: boolean;
  cancelled_by: string | null;
  attended: boolean;
  attended_time: string | null;
  ended_time: string | null;
  class_rating: string;
  class_review: string;
}

export type ParticipantsReadResponse = Participant;

export type ParticipantsListResponse = ListResponse<Participant>;

export type ParticipantsListRequestParams = Partial<{
  page: number;
  room_token: string;
  room_slug: string;
  company_slug: string;
  userid: string;
  username: string;
  attended: boolean;
  attended_time_before: string;
  attended_time_after: string;
  ended_time_before: string;
  ended_time_after: string;
  class_rating_lte: number;
  class_rating_gte: number;
  user_role: string;
  teacher: string;
}>;

export interface ParticipantsCreateRequestParams {
  room_token: string;
  room_slug?: string;
  company_slug?: string;
  teacher?: string;
  user_role: string;
  username: string;
  cancelled: boolean;
  attended_time: string;
  ended_time: string;
  class_rating: string;
  class_review: string;
}

export async function list(
  params: ParticipantsListRequestParams,
  token: string,
): Promise<ClientResponse<ParticipantsListResponse>> {
  return client.get(PARTICIPANTS_ENDPOINT, params, token);
}

export async function read(
  uuid: string,
  token: string,
): Promise<ClientResponse<ParticipantsReadResponse>> {
  return client.get(`${PARTICIPANTS_ENDPOINT}/${uuid}`, undefined, token);
}

export async function create(
  params: ParticipantsCreateRequestParams,
  token: string,
): Promise<ClientResponse<ParticipantsListResponse>> {
  return client.get(PARTICIPANTS_ENDPOINT, params, token);
}


src/client.ts
import fetch, { Response } from 'node-fetch';

export interface ErrorResponse {
  message: string;
  error_code: string;
}

export interface NotFoundErrorResponse {
  detail: 'Not found.';
}

export type BadRequestErrorResponse<R> = {
  [K in keyof R]?: string;
};

export class ApiClientError extends Error {
  constructor(
    public readonly message: string,
    public readonly errorCode?: string,
  ) {
    super(message);
  }
}

export type ClientResponse<T> =
  | {
      status: 200 | 201;
      body: T;
    }
  | {
      status: 400;
      body: BadRequestErrorResponse<{ [key: string]: string }>;
    }
  | {
      status: 404;
      body: NotFoundErrorResponse;
    }
  | {
      status: 204 | 405 | 500;
    };

export async function createClientResponse<T>(
  response: Response,
): Promise<ClientResponse<T>> {
  const status = response.status as 200 | 201;

  try {
    return Promise.resolve({
      status,
      body: (await response.json()) as T,
    });
  } catch (e) {
    console.info(e);
    return Promise.resolve({
      status: status as 204 | 405 | 500,
    });
  }
}

export async function get<P, R>(
  endpoint: string,
  params?: P,
  token?: string,
): Promise<ClientResponse<R>> {
  const urlSearchParams = new URLSearchParams(
    Object.entries(params || {}),
  ).toString();

  const url = new URL(
    urlSearchParams.length > 0
      ? `${endpoint}?${urlSearchParams.toString()}`
      : endpoint,
  );

  return request('GET', url.toString(), undefined, token);
}

export async function post<P, R>(
  endpoint: string,
  params: P,
  token?: string,
): Promise<ClientResponse<R>> {
  return request('POST', endpoint, params, token);
}

export async function put<P, R>(
  endpoint: string,
  params: P,
  token?: string,
): Promise<ClientResponse<R>> {
  return request('PUT', endpoint, params, token);
}

export async function del<P>(
  endpoint: string,
  params?: P,
  token?: string,
): Promise<ClientResponse<undefined>> {
  return request('DELETE', endpoint, params, token);
}

export async function request<P, R>(
  method: 'GET' | 'POST' | 'DELETE' | 'PUT',
  endpoint: string,
  params?: P,
  token?: string,
): Promise<ClientResponse<R>> {
  const headers: HeadersInit = {};
  let body;

  if (['PUT', 'POST'].includes(method)) {
    body = JSON.stringify(params);
    headers['Content-Type'] = 'application/json';
    headers['Content-Length'] = body.length.toString();
  }

  if (token) {
    headers['Authorization'] = `Bearer ${token}`;
  }

  const response = await fetch(endpoint, {
    method,
    headers,
    body,
  });

  return createClientResponse(response);
}
